docs/root/configuration/http/http_filters/header_mutation_filter.rst

```diff
@@ -0,0 +1,20 @@
+.. _config_http_filters_header_mutation:
+
+Header Mutation
+===============
+
+* This filter should be configured with the type URL ``type.googleapis.com/envoy.extensions.filters.http.header_mutation.v3.HeaderMutation``.
+* :ref:`v3 API reference <envoy_v3_api_msg_extensions.filters.http.header_mutation.v3.HeaderMutation>`
+
+This is a filter that can be used to add, remove, append, or update HTTP headers. It can be added in any position in the filter chain
+and used as downstream or upstream HTTP filter. The filter can be configured to apply the header mutations to the request, response, or both.
+
+
+In most cases, this filter would be a more flexible alternative to the ``request_headers_to_add``, ``request_headers_to_remove``,
+``response_headers_to_add``, and ``response_headers_to_remove`` fields in the :ref:`route configuration <envoy_v3_api_msg_config.route.v3.RouteConfiguration>`.
+Because the filter provides complete control over the position and order of the header mutations and it can be used as upstream filter and
+mutates the request headers after load balancing and host selection.
+
+
```

Comment:
One of the interesting properties of this design, which is probably worth spelling out, is that if a later filter clears the route cache, the mutated headers will influence the new route pick. That was never the case for `request_headers_to_add` etc. I think this makes the new filter quite a bit more useful in some configurations.

Reply:
This is oneof the target scenes. (And I think may be we can add a `clear_route_cache` option in the filter config in the future? 🤔

Reply:
"Header mutations may be used to influence later route picks if the route cache is cleared by a filter executing after the header mutation filter"

Yeah, I think a clear_route_cache option could make sense when there is a use case. I'd split it to a different PR.